core/textile/backup.go

```diff
@@ -0,0 +1,266 @@
+package textile
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+
+	"github.com/FleekHQ/space-daemon/log"
+
+	"github.com/textileio/go-threads/core/thread"
+)
+
+// backup a single file with an io.Reader provided
+func (tc *textileClient) BackupFileWithReader(ctx context.Context, bucket Bucket, path string, reader io.Reader) (err error) {
+
+	if _, _, err = tc.UploadFileToHub(ctx, bucket, path, reader); err != nil {
+		log.Error(fmt.Sprintf("error backuping up file with reader (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+		return err
+	}
+
+	if _, err = tc.setMirrorFileBackup(ctx, path, bucket.Slug()); err != nil {
+		log.Error(fmt.Sprintf("error setting mirror file as backup (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+		return err
+	}
+
+	return nil
+}
+
+// backup a single file with a path provided
+func (tc *textileClient) backupFile(ctx context.Context, bucket Bucket, path string) error {
+
+	errc := make(chan error, 1)
+	pipeReader, pipeWriter := io.Pipe()
+
+	// go routine for piping
+	go func() {
+		defer close(errc)
+		defer pipeWriter.Close()
+
+		err := bucket.GetFile(ctx, path, pipeWriter)
+		if err != nil {
+			errc <- err
+			log.Error(fmt.Sprintf("error getting file (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+			return
+		}
+	}()
+	if err := <-errc; err != nil {
+		return err
+	}
+
+	if err := tc.BackupFileWithReader(ctx, bucket, path, pipeReader); err != nil {
+		log.Error(fmt.Sprintf("error backing up file (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+		return err
+	}
+
+	return nil
+}
+
+// backup all files in a bucket
+func (tc *textileClient) backupBucketFiles(ctx context.Context, bucket Bucket, path string) (int, error) {
+	var wg sync.WaitGroup
+	var count int
+
+	// XXX: we ignore errc (no atomicity at all) for now but we should return
```

Comment:
Maybe we should think through a more atomic/consistent way of doing backups, for example if something goes wrong during a backup, do we plan to start all over again or we should actually track what failed and only backup those?

Reply:
@perfectmak 

I don't know. I don't want to delay this issue any longer but I agree we should decide on atomicity. What do you think, @jsonsivar @dmerrill6 ?

Reply:
It would be nicer to not have to start all over again. So maybe there is a backup state (not started, in progress, up-to-date) that we can set and sync based off.

Reply:
@jsonsivar 

We're keeping the successful state via `setMirrorFileBackup` and `unsetMirrorFileBackup`. I've added `isMirrorBackupFile` in `backupBucketFiles` in 7cb351cedbc82052b2cdbba91c391b277c9b2160 to avoid starting all over again (waiting for the hub to recover so I can test it), that should be good for now.